src/WebJobs.Script/Extensions/HttpRequestExtensions.cs

```diff
@@ -102,6 +104,38 @@ public static bool IsPlatformInternalRequest(this HttpRequest request, IEnvironm
             return string.Compare(value, "true", StringComparison.OrdinalIgnoreCase) == 0;
         }
 
+        public static bool IsInternalAuthAllowed(this HttpRequest httpRequest)
+        {
+            // First verify that this is an internal request
+            if (!httpRequest.IsAppServiceInternalRequest())
+            {
+                return false;
+            }
+
+            // Next check to see if the specific API is allowed based on whitelist
+            var options = httpRequest.HttpContext.RequestServices.GetService<IOptions<FunctionsHostingConfigOptions>>();
+            string whitelist = options?.Value.InternalAuthAllowedApiWhitelist;
+            if (whitelist != null)
+            {
+                // A whitelist is configured, so we ensure that the current request
+                // matches any of the whitelisted APIs.
+                var whitelistedAPIs = whitelist.Split(new[] { '|' }, StringSplitOptions.RemoveEmptyEntries).ToHashSet(StringComparer.OrdinalIgnoreCase);
```

Comment:
Can we move this logic to the options setup so this is only run once, when the option is retrieved, and the results are assigned and cached with the options?

Reply:
I added some on demand caching in the options class